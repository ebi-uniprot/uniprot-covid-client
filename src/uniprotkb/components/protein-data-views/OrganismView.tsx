import React, { Fragment } from 'react';
import { InfoList, ExternalLink } from 'franklin-sites';
import { OrganismData } from '../../adapters/namesAndTaxonomyConverter';
import UniProtKBEvidenceTag from './UniProtKBEvidenceTag';
import externalUrls from '../../config/externalUrls';

type OrganismDataProps = {
  data: OrganismData;
};

export const OrganismLineage: React.FC<{ lineage: string[] }> = ({
  lineage,
}) => <Fragment>{lineage.join(' > ')}</Fragment>;

export const OrganismId: React.FC<{ taxonId: number | undefined }> = ({
  taxonId,
}) => {
  if (!taxonId) {
    return null;
  }
  return (
    <ExternalLink url={externalUrls.NCBI(taxonId)}>{taxonId}</ExternalLink>
  );
};

const OrganismView: React.FC<OrganismDataProps> = ({
  data,
}): JSX.Element | null => {
  if (!data) {
    return null;
  }

  const termValue = `${data.scientificName}${
    data.commonName ? ` (${data.commonName})` : ''
  } ${data.synonyms && data.synonyms.length > 0 ? ` (${data.synonyms})` : ''}`;

  return (
    <ExternalLink url={externalUrls.NCBI(data.taxonId)}>
      {termValue}
    </ExternalLink>
  );
};

export const OrganismListView: React.FC<{
  data?: OrganismData;
  hosts?: OrganismData[];
}> = ({ data, hosts }): JSX.Element | null => {
  if (!data) {
    return null;
  }
  const infoListData: { title: string; content: JSX.Element | string }[] = [];
  if (data.scientificName) {
    infoListData.push({
      title: 'Organism',
      content: (
        <Fragment>
          {data.taxonId && (
            <ExternalLink url={externalUrls.NCBI(data.taxonId)}>
              {`${data.scientificName} (${data.commonName})`}
            </ExternalLink>
          )}
          {data.evidences && data.evidences.length && (
            <UniProtKBEvidenceTag evidences={data.evidences} />
          )}
        </Fragment>
      ),
    });
  }
  if (data.taxonId) {
    infoListData.push({
      title: 'Taxonomic identifier',
      content: <OrganismId taxonId={data.taxonId} />,
    });
  }
  if (data.lineage) {
    infoListData.push({
      title: 'Taxonomic lineage',
      content: <OrganismLineage lineage={data.lineage} />,
    });
  }
  if (hosts) {
    infoListData.push({
      title: 'Virus hosts',
      content: (
        <Fragment>
          {hosts.map((host) => (
            <p key={host.taxonId}>
              <OrganismView data={host} />
            </p>
          ))}
        </Fragment>
      ),
    });
  }
  return <InfoList infoData={infoListData} />;
};

export default OrganismView;
