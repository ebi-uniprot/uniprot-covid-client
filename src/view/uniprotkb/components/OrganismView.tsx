import React, { Fragment } from 'react';
import { InfoList, ExternalLink } from 'franklin-sites';
import { OrganismData } from '../../../model/uniprotkb/sections/NamesAndTaxonomyConverter';
import UniProtEvidenceTag from '../../../components/UniProtEvidenceTag';
import externalUrls from '../../../utils/externalUrls';

type OrganismDataProps = {
  data: OrganismData;
};

export const OrganismLineage: React.FC<{ lineage: string[] }> = ({
  lineage,
}) => <Fragment>{lineage.join(' > ')}</Fragment>;

export const OrganismId: React.FC<{ taxonId: number | undefined }> = ({
  taxonId,
}) => {
  if (!taxonId) {
    return null;
  }
  return (
    <Fragment>
      <a href={`//www.uniprot.org/taxonomy/${taxonId}`}>{`${taxonId} `}</a>
      <ExternalLink url={externalUrls.NCBI(taxonId)}>NCBI</ExternalLink>
    </Fragment>
  );
};

const OrganismView: React.FC<OrganismDataProps> = ({
  data,
}): JSX.Element | null => {
  if (!data) {
    return null;
  }

  const termValue = `${data.scientificName}${
    data.commonName ? ` (${data.commonName})` : ''
  } ${data.synonyms && data.synonyms.length > 0 ? ` (${data.synonyms})` : ''}`;

  return (
    <a href={`https://www.uniprot.org/taxonomy/${data.taxonId}`}>{termValue}</a>
  );
};

export const OrganismListView: React.FC<{
  data?: OrganismData;
  hosts?: OrganismData[];
}> = ({ data, hosts }): JSX.Element | null => {
  if (!data) {
    return null;
  }
  const infoListData: { title: string; content: JSX.Element | string }[] = [];
  if (data.scientificName) {
    infoListData.push({
      title: 'Organism',
      content: (
        <Fragment>
          <a href={`//www.uniprot.org/taxonomy/${data.taxonId}`}>
            {`${data.scientificName} (${data.commonName})`}
          </a>
          {data.evidences && data.evidences.length && (
            <UniProtEvidenceTag evidences={data.evidences} />
          )}
        </Fragment>
      ),
    });
  }
  if (data.taxonId) {
    infoListData.push({
      title: 'Taxonomic identifier',
      content: <OrganismId taxonId={data.taxonId} />,
    });
  }
  if (data.lineage) {
    infoListData.push({
      title: 'Taxonomic lineage',
      content: <OrganismLineage lineage={data.lineage} />,
    });
  }
  if (hosts) {
    infoListData.push({
      title: 'Virus hosts',
      content: (
        <Fragment>
          {hosts.map(host => (
            <p key={host.taxonId}>
              <OrganismView data={host} />
            </p>
          ))}
        </Fragment>
      ),
    });
  }
  return <InfoList infoData={infoListData} />;
};

export default OrganismView;
