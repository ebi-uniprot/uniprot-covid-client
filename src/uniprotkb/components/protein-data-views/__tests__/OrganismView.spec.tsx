import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { render } from '@testing-library/react';
import OrganismView from '../OrganismView';
import OrganismUIDataJson from './__mocks__/organismUIData.json';

describe('Organism', () => {
  test('should render organism', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <OrganismView data={OrganismUIDataJson} />
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
