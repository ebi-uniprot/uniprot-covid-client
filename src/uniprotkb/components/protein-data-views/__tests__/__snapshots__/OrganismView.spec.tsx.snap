// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Organism should render organism 1`] = `
<DocumentFragment>
  <a
    class="external-link"
    href="https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?lvl=0&id=12344"
    rel="noopener"
    target="_blank"
  >
    Cheeky monkey (Really cheeky little monkey)  (Monkey 1,Monkey 2)
    <svg
      fill="currentColor"
      viewBox="0 0 576 512"
      width="12.5"
    >
      <path
        d="M576 24v128c0 21-26 32-41 17l-36-36-243 244c-10 9-25 9-34 0l-23-23c-9-9-9-24 0-34L443 77l-36-36c-15-15-4-41 17-41h128c13 0 24 11 24 24zM407 271l-16 16a24 24 0 00-7 17v144H64V128h264a24 24 0 0017-7l16-16c15-15 4-41-17-41H48C21 64 0 85 0 112v352c0 27 21 48 48 48h352c27 0 48-21 48-48V288c0-22-26-32-41-17z"
      />
    </svg>
  </a>
</DocumentFragment>
`;
